daemon/logger/loggerutils/logfile.go

```diff
@@ -2,42 +2,104 @@ package loggerutils // import "github.com/docker/docker/daemon/logger/loggerutil
 
 import (
 	"bytes"
+	"compress/gzip"
 	"context"
+	"encoding/json"
 	"fmt"
 	"io"
 	"os"
 	"strconv"
+	"strings"
 	"sync"
 	"time"
 
 	"github.com/docker/docker/daemon/logger"
 	"github.com/docker/docker/daemon/logger/loggerutils/multireader"
 	"github.com/docker/docker/pkg/filenotify"
+	"github.com/docker/docker/pkg/pools"
 	"github.com/docker/docker/pkg/pubsub"
 	"github.com/docker/docker/pkg/tailfile"
 	"github.com/fsnotify/fsnotify"
 	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
 )
 
+const tmpLogfileSuffix = ".tmp"
+
+// rotateFileMetadata is a metadata of the gzip header of the compressed log file
+type rotateFileMetadata struct {
+	LastTime time.Time `json:"lastTime,omitempty"`
+}
+
+// refCounter is a counter of logfile being referenced
+type refCounter struct {
+	mu      sync.Mutex
+	counter map[string]int
+}
+
+// Reference increase the reference counter for specified logfile
+func (rc *refCounter) GetReference(fileName string, openRefFile func(fileName string, exists bool) (*os.File, error)) (*os.File, error) {
+	rc.mu.Lock()
+	defer rc.mu.Unlock()
+
+	var (
+		file *os.File
+		err  error
+	)
+	_, ok := rc.counter[fileName]
+	file, err = openRefFile(fileName, ok)
+	if err != nil {
+		return nil, err
+	}
+
+	if ok {
+		rc.counter[fileName]++
+	} else if file != nil {
+		rc.counter[file.Name()] = 1
+	}
+
+	return file, nil
+}
+
+// Dereference reduce the reference counter for specified logfile
+func (rc *refCounter) Dereference(fileName string) error {
+	rc.mu.Lock()
+	defer rc.mu.Unlock()
+
+	rc.counter[fileName]--
+	if rc.counter[fileName] <= 0 {
+		delete(rc.counter, fileName)
+		err := os.Remove(fileName)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
 // LogFile is Logger implementation for default Docker logging.
 type LogFile struct {
-	f             *os.File // store for closing
-	closed        bool
-	mu            sync.RWMutex
-	capacity      int64 //maximum size of each file
-	currentSize   int64 // current size of the latest file
-	maxFiles      int   //maximum number of files
-	notifyRotate  *pubsub.Publisher
-	marshal       logger.MarshalFunc
-	createDecoder makeDecoderFunc
-	perms         os.FileMode
+	logPath         string
```

Comment:
Actually, given that `LogFile.LogPath()` seems to be unused, it looks like it's only used in `checkCapacityAndRotate()`. Can't we just use `f.Name()`?

Something like:

```patch
diff --git a/daemon/logger/loggerutils/logfile.go b/daemon/logger/loggerutils/logfile.go
index e0f21f46c..428f4786b 100644
--- a/daemon/logger/loggerutils/logfile.go
+++ b/daemon/logger/loggerutils/logfile.go
@@ -79,7 +79,6 @@ func (rc *refCounter) Dereference(fileName string) error {
 
 // LogFile is Logger implementation for default Docker logging.
 type LogFile struct {
-	logPath         string
 	mu              sync.RWMutex // protects the logfile access
 	f               *os.File     // store for closing
 	closed          bool
@@ -111,7 +110,6 @@ func NewLogFile(logPath string, capacity int64, maxFiles int, compress bool, mar
 	}
 
 	return &LogFile{
-		logPath:         logPath,
 		f:               log,
 		capacity:        capacity,
 		currentSize:     size,
@@ -162,15 +160,16 @@ func (w *LogFile) checkCapacityAndRotate() error {
 
 	if w.currentSize >= w.capacity {
 		w.rotateMu.Lock()
+		fname := w.f.Name()
 		if err := w.f.Close(); err != nil {
 			w.rotateMu.Unlock()
 			return errors.Wrap(err, "error closing file")
 		}
-		if err := rotate(w.logPath, w.maxFiles, w.compress); err != nil {
+		if err := rotate(fname, w.maxFiles, w.compress); err != nil {
 			w.rotateMu.Unlock()
 			return err
 		}
-		file, err := os.OpenFile(w.logPath, os.O_WRONLY|os.O_TRUNC|os.O_CREATE, w.perms)
+		file, err := os.OpenFile(fname, os.O_WRONLY|os.O_TRUNC|os.O_CREATE, w.perms)
 		if err != nil {
 			w.rotateMu.Unlock()
 			return err
@@ -185,7 +184,7 @@ func (w *LogFile) checkCapacityAndRotate() error {
 		}
 
 		go func() {
-			compressFile(w.logPath+".1", w.lastTimestamp)
+			compressFile(fname+".1", w.lastTimestamp)
 			w.rotateMu.Unlock()
 		}()
 	}
@@ -262,11 +261,6 @@ func compressFile(fileName string, lastTimestamp time.Time) {
 	}
 }
 
-// LogPath returns the location the given writer logs to.
-func (w *LogFile) LogPath() string {
-	return w.logPath
-}
-
 // MaxFiles return maximum number of files
 func (w *LogFile) MaxFiles() int {
 	return w.maxFiles
```


Or possibly even:

```patch
diff --git a/daemon/logger/loggerutils/logfile.go b/daemon/logger/loggerutils/logfile.go
index e0f21f46c..5ad023423 100644
--- a/daemon/logger/loggerutils/logfile.go
+++ b/daemon/logger/loggerutils/logfile.go
@@ -79,7 +79,6 @@ func (rc *refCounter) Dereference(fileName string) error {
 
 // LogFile is Logger implementation for default Docker logging.
 type LogFile struct {
-	logPath         string
 	mu              sync.RWMutex // protects the logfile access
 	f               *os.File     // store for closing
 	closed          bool
@@ -111,7 +110,6 @@ func NewLogFile(logPath string, capacity int64, maxFiles int, compress bool, mar
 	}
 
 	return &LogFile{
-		logPath:         logPath,
 		f:               log,
 		capacity:        capacity,
 		currentSize:     size,
@@ -166,11 +164,11 @@ func (w *LogFile) checkCapacityAndRotate() error {
 			w.rotateMu.Unlock()
 			return errors.Wrap(err, "error closing file")
 		}
-		if err := rotate(w.logPath, w.maxFiles, w.compress); err != nil {
+		if err := rotate(w.f.Name(), w.maxFiles, w.compress); err != nil {
 			w.rotateMu.Unlock()
 			return err
 		}
-		file, err := os.OpenFile(w.logPath, os.O_WRONLY|os.O_TRUNC|os.O_CREATE, w.perms)
+		file, err := os.OpenFile(w.f.Name(), os.O_WRONLY|os.O_TRUNC|os.O_CREATE, w.perms)
 		if err != nil {
 			w.rotateMu.Unlock()
 			return err
@@ -185,7 +183,7 @@ func (w *LogFile) checkCapacityAndRotate() error {
 		}
 
 		go func() {
-			compressFile(w.logPath+".1", w.lastTimestamp)
+			compressFile(w.f.Name()+".1", w.lastTimestamp)
 			w.rotateMu.Unlock()
 		}()
 	}
@@ -262,11 +260,6 @@ func compressFile(fileName string, lastTimestamp time.Time) {
 	}
 }
 
-// LogPath returns the location the given writer logs to.
-func (w *LogFile) LogPath() string {
-	return w.logPath
-}
-
 // MaxFiles return maximum number of files
 func (w *LogFile) MaxFiles() int {
 	return w.maxFiles
```

Reply:
It's nice to store it because `w.f.Name()` can be racey  (since `w.f` will change) unless locking is involved meanwhile the actual path never changes.

Reply:
Yes, I was looking at that, seems the only place `LogFile.f` is written to is during construction in  [`NewLogFile()`](https://github.com/moby/moby/blob/241c904e6f5fff020890a7641558e83a209c0bbd/daemon/logger/loggerutils/logfile.go#L52), and in [`checkCapacityAndRotate()`](https://github.com/moby/moby/blob/241c904e6f5fff020890a7641558e83a209c0bbd/daemon/logger/loggerutils/logfile.go#L109), and the filename never actually changes (it's always `/var/lib/docker/containers/<container-id>/<container-id>-json.log`).

Perhaps the first option, using an intermediate variable (`fname := w.f.Name()`) would work then.

Reply:
I think what @thaJeztah said makes sense. @cpuguy83 WDYT?

Reply:
This works, but still need to lock to get the file name.

Reply:
@cpuguy83  Yeah, we can get the file name after `w.rotateMu.Lock()`.